turbopack/crates/turbopack-core/src/issue/mod.rs

```diff
@@ -512,6 +540,66 @@ impl IssueSource {
     }
 }
 
+async fn source_pos(
+    source: Vc<Box<dyn Source>>,
+    origin: Vc<FileSystemPath>,
+    start: SourcePos,
+    end: SourcePos,
+) -> Result<Option<(Vc<Box<dyn Source>>, SourcePos, SourcePos)>> {
+    let Some(generator) = Vc::try_resolve_sidecast::<Box<dyn GenerateSourceMap>>(source).await?
+    else {
+        return Ok(None);
+    };
+
+    let srcmap = generator.generate_source_map();
+
+    let Some(srcmap) = *convert_to_turbopack_source_map(srcmap, origin).await? else {
+        return Ok(None);
+    };
+
+    let find = |line: usize, col: usize| async move {
+        let TokenWithSource {
+            token,
+            source_content,
+        } = &*srcmap.lookup_token_and_source(line, col).await?;
+
+        match &*token.await? {
+            crate::source_map::Token::Synthetic(t) => Ok::<_, anyhow::Error>((
+                SourcePos {
+                    line: t.generated_line as _,
+                    column: t.generated_column as _,
+                },
+                *source_content,
+            )),
+            crate::source_map::Token::Original(t) => Ok((
+                SourcePos {
+                    line: t.original_line as _,
+                    column: t.original_column as _,
+                },
+                *source_content,
+            )),
+        }
+    };
+
+    let (start, content_1) = find(start.line, start.column).await?;
+    let (end, content_2) = find(end.line, end.column).await?;
+
+    let content = content_1.or(content_2);
+
+    let source = if let Some(content) = content {
+        let file = File::from(content.await?.clone_value());
+
+        let content = FileContent::new(file).cell();
+        let content = AssetContent::file(content);
+
+        Vc::upcast(VirtualSource::new(source.ident().path(), content))
+    } else {
+        source
+    };
+
+    Ok(Some((source, start, end)))
```

Comment:
This can report some invalid locations in some cases. We never want to report a source with a line/column that doesn't exist there or doesn't map to that.

1. If `content` is `None`, we shouldn't report `source` with `start` and `end` since these are not valid locations there
2. If `content_1 != content_2` we can't report `end` with `content_1` since that's not a valid location there